design/client-policies.md

```diff
@@ -0,0 +1,191 @@
+
+* Enforce more secure client authentication method when client registration
+* Enforce Holder-of-Key Token
+* Enforce Proof Key for Code Exchange (PKCE)
+* Enforce secure signature algorithm for Signed JWT client authentication (private-key-jwt)
+* Enforce HTTPS redirect URI
+* Enforce Request Object satisfying high security level
+* Enforce Response Type of OIDC Hybrid Flow
+* Enforce more secure state and nonce treatment for preventing CSRF
+* Enforce more secure signature algorithm when client registration
+
+Also, one of several purposes for this executor is to take over [Client Registration Policy](https://www.keycloak.org/docs/latest/securing_apps/index.html#client-registration-policies).
+
+
+### Policy : how to determine which client and what action
+
+A policy consists of several conditions and executors. A policy can be adpoted to a client satisfying all conditions of this policy. All executors execute their task against this client.
+
+Considering both a condition and an executor do their task independently from other conditions and other executors, the administrator can not specify in which order conditions and executors are evaluated.
+
+#### Pre-set Policy
+
+A pre-set policy is a client policy that have registered in keycloak as default.
+This pre-set policy can be used to enforce basic security level to all clients.
+
+### Endpoint and Opeation : when and where policy is adopted
+
+Conditions and executors of a policy should be evaluated on endpoints where a client application interact with keycloak by OAuth2/OIDC protocol. Considering this point, when and where a policy can be adopted to a client is as follows:
+
+##### OIDC Dynamic Client Registration
+  * Opeation
+    * Register
+    * Update
+  * Authentication
+    * Authenticated (Using Initial Access Token, Registration Access Token)
+    * Anonymous
+
+##### Admin REST API
+  * Opeation
+    * Register
+    * Update
+
+##### Authorization Endpoint
+  * Operation
+    * Authorization Request
+
+##### Token Endpoint
+  * Operation
+    * Token Request
+    * Token Refresh Request
+
+##### Logout Endpoint
+  * Operation
+    * Logout Request by Refresh Token
+
+##### Token Introspection Endpoint
+  * Operation
+    * Token Request
+
+##### UserInfo Endpoint
+  * Operation
+    * UserInfo Request
+
+### Configuration
+
+Policies, conditions, executors can be configured on Admin Console. To do so, add "Client Policies" tab on Realm->Realm Settings, which means an administrator can have client policies per realm.
+
+An administrator can create new policy and add conditions and executors.
+An administrator can update their created policies by add/modify/delete conditions and executors while the administrator can not update pre-set policies.
+
+### Backward Compatibility
+
+Client Policies can take over Client Registration. However, Client Registration Policies also still co-exist. In the future, Client Registration Policies will be removed.
```

Comment:
Ideal would be to somehow automatically migrate client registration policies into client policies so we can get rid of them

Reply:
I've missed this point. I'll add this automatic migration to Backward Compatibility section.